processing/src/main/java/org/apache/druid/segment/incremental/OakIncrementalIndex.java

```diff
@@ -0,0 +1,573 @@
+    }
+
+
+    public Iterator<Row> transformIterator(boolean descending, Function<Map.Entry<OakRBuffer, OakRBuffer>, Row> transformer)
+    {
+      OakMap<IncrementalIndexRow, Row> tmpOakMap = descending ? oak.descendingMap() : oak;
+      return tmpOakMap.zc().entrySet().stream().map(transformer).iterator();
+    }
+
+
+    private IncrementalIndexRow getMinIncrementalIndexRow()
+    {
+      return new IncrementalIndexRow(minTimestamp, null, dimensionDescsList, IncrementalIndexRow.EMPTY_ROW_INDEX);
+    }
+
+    @Override
+    public int getPriorIndex(IncrementalIndexRow key)
+    {
+      return 0;
+    }
+
+    @Override
+    public long getMinTimeMillis()
+    {
+      return oak.firstKey().getTimestamp();
+    }
+
+    @Override
+    public long getMaxTimeMillis()
+    {
+      return oak.lastKey().getTimestamp();
+    }
+
+    @Override
+    public Iterator<IncrementalIndexRow> iterator(boolean descending)
+    {
+      // We should never get here because we override iterableWithPostAggregations
+      throw new NotImplementedException();
+    }
+
+    @Override
+    public Iterable<IncrementalIndexRow> timeRangeIterable(boolean descending, long timeStart, long timeEnd)
+    {
+      if (timeStart > timeEnd) {
+        return null;
+      }
+      return () -> {
+        IncrementalIndexRow from = new IncrementalIndexRow(timeStart, null, dimensionDescsList, IncrementalIndexRow.EMPTY_ROW_INDEX);
+        IncrementalIndexRow to = new IncrementalIndexRow(timeEnd, null, dimensionDescsList, IncrementalIndexRow.EMPTY_ROW_INDEX);
+
+        try (OakMap<IncrementalIndexRow, Row> subMap = oak.subMap(from, true, to, false, descending)) {
+          Iterator<Map.Entry<OakRBuffer, OakRBuffer>> iterator = subMap
+                  .zc()
+                  .entrySet()
+                  .iterator();
+          //TODO YONIGO - what sorcery is this transform function doing to allow this?!
+          return Iterators.transform(iterator,
+              entry -> new OakIncrementalIndexRow(entry.getKey(), dimensionDescsList, entry.getValue()));
+        }
+      };
+    }
+
+    @Override
+    public Iterable<IncrementalIndexRow> keySet()
+    {
+      return () -> Iterators.transform(oak.zc().entrySet().iterator(), entry ->
+              new OakIncrementalIndexRow(entry.getKey(), dimensionDescsList, entry.getValue()));
+    }
+
+    @Override
+    public Iterable<IncrementalIndexRow> persistIterable()
+    {
+      return keySet();
+    }
+
+    @Override
+    public int putIfAbsent(IncrementalIndexRow key, int rowIndex)
+    {
+      //Oak is pigibacking the FactsHolder and doesnt really have rowIndex stored in it.
+      throw new NotImplementedException();
+    }
+
+    @Override
+    public void clear()
+    {
+      oak.clear();
+    }
+
+    private AddToFactsResult addToOak(
+            InputRow row,
+            AtomicInteger numEntries,
+            AtomicLong bytesInMemory, IncrementalIndexRow incrementalIndexRow,
```

Comment:
If we're not using `bytesInMemory` here, then we don't need to pass it along 
```suggestion
            IncrementalIndexRow incrementalIndexRow,
```